Implement reading extra inhabitants of integers

augusto2112 opened the pull request:
Bools are special in that they only need 1 bit of storage, and have many extra inhabitants available. This adds a test to check that an optional whose payload is a bool is reflected correctly.

augusto2112 committed:
Add the reflect_Optional_Bool validation test

Bools are special in that they only need 1 bit of storage, and have many
extra inhabitants available. This adds a test to check that an optional
whose payload is a bool is reflected correctly.

augusto2112 commented:
@swift-ci smoke test

tbkka commented on validation-test/Reflection/reflect_Optional_Bool.swift:
```diff
@@ -0,0 +1,35 @@
+import SwiftReflectionTest
+
+let value: Bool? = false
+reflect(enum: value)
```
Since there are only three values -- true, false, nil -- I would copy this three times so we exercise every possibility.

We also want to check the _value_, not just the type.

mikeash replied:
Agreed. While we're at it, do we also want to check `Bool??` just to be sure?

augusto2112 replied:
@tbkka good call! There *is* a bug when reflecting optional bools after all. Both `true` and `false` reflect as nil:

```
Enum value:
(enum_value name=none index=1)
```

And nil fails:

```
Enum value:
swift_reflection_projectEnumValue failed.
```

This is most likely the bug I'm seeing in LLDB as well.

tbkka replied:
Huzzah!  Once you have a test case that reproduces the bug, the rest should be easy!  😁

mikeash commented on validation-test/Reflection/reflect_Optional_Bool.swift:
```diff
@@ -0,0 +1,35 @@
+// RUN: %target-build-swift -g -lswiftSwiftReflectionTest %s -o %t/reflect_Optional_Bool
+// RUN: %target-codesign %t/reflect_Optional_Bool
+
+// RUN: %target-run %target-swift-reflection-test %t/reflect_Optional_Bool | %FileCheck %s --check-prefix=CHECK-%target-ptrsize
```
Nothing in this output should depend on the pointer size, so you should delete this `--check-prefix` flag, and change all the check comments to just `// CHECK:`. That should let it work seamlessly on 32-bit as well.

mikeash reviewed (changes requested):
Thanks for adding this!

augusto2112 committed:
Implement reading extra inhabitants of integers

Bools may have extra inhabitants. Implement reading the extra
inhabitants of bools in BuiltinTypeInfo::readExtraInhabitantIndex.

augusto2112 commented:
@swift-ci smoke test

mikeash commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -241,25 +241,50 @@ BuiltinTypeInfo::BuiltinTypeInfo(TypeRefBuilder &builder,
+      if (!reader.readInteger(address, getSize(), &rawValue))
+        return false;
+
+      auto maxRawValue = (256 << (getSize() - 1)) - 1;
```
This calculation will overflow if `getSize()` is greater than `4`. Not an issue if the above code bails out when `getSize()` is too big. Or this can do the math in a `uint64_t` to avoid overflow.

mikeash commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -510,7 +535,7 @@ class SinglePayloadEnumTypeInfo: public EnumTypeInfo {
       int XITag;
       if (!PayloadCase.TI.readExtraInhabitantIndex(reader, address, &XITag)) {
-        return false;
+        return false;;
```
Stray semicolon.

mikeash commented on validation-test/Reflection/reflect_Optional_Bool.swift:
```diff
@@ -0,0 +1,208 @@
+// CHECK:   (struct Swift.Bool))
+// CHECK: )
+
+let optOptNil: Bool?? = .none
```
Nice comprehensive test!

augusto2112 commented:
@swift-ci smoke test

mikeash reviewed (approved):
Restricting it to `Bool` simplifies things quite a bit! LGTM.
